Flatten nested optionals resulting from try? and optional sub-expressions

bjhomer opened the pull request:
This PR implements the flattening of a double-optional type produced from code like `let x = try? optionalThing()`. The pitch thread can be found here: https://forums.swift.org/t/make-try-optional-chain-flattening-work-together/7415/55

bjhomer committed:
Make 'try?' flatten optional chaining like optional-chaining does.

If the sub-expression of the 'try?' is optional, the result will be the same level of optional-ness.
If the sub-expression is non-optional, the result is optional.

Thus, the following lines all end up with the same type of 'Int?'
 - let x = try? 3 as Int
 - let x = try? 3 as? Int
 - let x = try? 3 as Int?

bjhomer committed:
Adjust SILGen tests for flattened optional try.

bjhomer committed:
Add backwards-compatible behavior for swift-version < 5

bjhomer committed:
Add a migration for the Swift 5 'try?' changes.

This migrates anything using the following pattern:

```swift
if let optX = try? someOptional(),
    let x = optX,
    ...
```

to this:

```swift
if let x = try? someOptional(),
    ...
```

bjhomer committed:
Run tests for the new 'try?' behavior with '-swift-version 5'

bjhomer commented:
@brentdax Rebased on master, tests are passing locally. Could you run the compatibility suite tests?

brentdax commented:
@swift-ci please test source compatibility

rudkx reviewed (changes requested):
This change would benefit from additional test cases.

For example, what happens with something like:

```
func foo() throws -> Int? { return 1 }
let x = (try? foo())!!
```

With the new semantics I would expect this to fail type checking, hopefully with a good diagnostic.

It's possible we already have some interesting test cases along these lines, but they are in files not currently compiled with `-swift-version 5`. Can you take a look and add some more comprehensive tests to a file that is compiled with `-swift-version 5`?

bjhomer commented:
Yep, I'll get on this.

xedin reviewed (commented):
I agree with @rudkx, it would be great to add more tests for this, otherwise changes to Sema LGTM!

bjhomer committed:
Add comment to clarify Swift 4/5 behavioral difference for 'try?'

bjhomer committed:
Add more 'try?' tests.

bjhomer committed:
Add a bunch more 'try' tests.

Some of these are currently failing. For example:

```swift
// expected-error {{cannot convert value of type 'Int??' to specified type 'Int'}}
let _: Int = try? producer.produceDoubleOptionalInt()
```

This actually produces this warning:

```
value of optional type 'Int?' not unwrapped; did you mean to use 'try!' or chain with '?'?
```

Note that the value in question is not an `Int?`, so a suggestion of `try!` here is inappropriate.
(And, due to the change in semantics of 'try?', wouldn't change anything anyway.)

bjhomer commented:
I have added a bunch more tests. A couple of them are currently failing; it appears I need to make some changes for cases where we used to suggest changing `try?` to `try!`.

bjhomer committed:
Don't consider fixups that won't do anything.

Since 'try?' no longer unconditionally adds a layer of optional, converting it
to 'try!' will no longer unconditionally remove a layer of optional. So let's not
suggest it. This leads to better diagnostics anyway.

bjhomer commented:
@xedin, @rudkx Tests have been added. I discovered that a fix-it would suggest changing `try?` into `try!` in cases where that would no longer fix anything, so that has also been addressed. I think we're ready for further review.

rudkx commented:
@swift-ci Please smoke test

rudkx commented on test/Parse/try.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-typecheck-verify-swift
+// RUN: %target-typecheck-verify-swift -swift-version 5
```
Can you split this into two tests, one for ambient-version mode (with no changes to the existing test) and one for `-swift-version 5`?

We still want to test the older behavior as long as it is supported.

bjhomer replied:
Yeah, I can do that.

rudkx commented:
@nkcsgexi I cannot seem to currently add you as a reviewer. Can you take a look at the migrator changes?

nkcsgexi commented:
hmm, interesting. I've added myself as a reviewer and will take a look.

rudkx reviewed (changes requested):
I looked at the AST & Sema changes as well as the tests. LGTM. Once you split out the new tests into a separate file for `-swift-version 5` I think that portion is good to go.

bjhomer commented:
Regarding the migration, I have a test file that I've been using locally. I would be happy to add it as an automated test, if I could figure out how. Is there some example of how to run a test like that in the repository right now?

bjhomer committed:
Add test validating that we preserve existing Swift 4 'try?' behavior.

rudkx commented:
@nkcsgexi Can you answer the question re: migration testing?

rudkx commented:
@swift-ci Please smoke test

nkcsgexi commented:
@bjhomer you can follow some examples in test/migrator. Taking `remove_override.swift` as an example, it's both test driver and the code example before migration. After running the migrator, the result should be identical to `remove_override.swift.expected`.

bjhomer committed:
Add basic tests for 'try?' migration.

bjhomer commented:
@nkcsgexi: Thanks for that example. I'm bulking up these tests, and it's pointing out a few things I need to improve.

bjhomer committed:
Make the try? migration simpler and more reliable.

We were trying to get fancy rewriting cases where multiple 'if let' clauses were
used to unwrap a double optional, but in testing that turned out to be too
unreliable. Now we simply detect any 'try?' expressions whose behavior will change,
and add an explicit 'as OldType' to the expression, in order to preserve the original
behavior.

bjhomer commented:
The migration tests are ready to go. I've simplified the approach a bit after discovering some problems with the earlier one.

bjhomer commented on test/Parse/try_swift4.swift:
```diff
@@ -0,0 +1,260 @@
+// RUN: %target-typecheck-verify-swift -swift-version 4
```
Even though git is showing this as a new file, this just the existing `try.swift` test with a few new tests added to differentiate between the swift4 and the swift 5 behavior..

bjhomer committed:
Rename try_swift4.swift back to try.swift.

It makes it clearer what has changed.

tkremenek commented:
This would be good to land before Swift 5 branches one last time from `master` on November 16.

bjhomer commented:
Agreed. I think everything is ready to go, but if there's any further feedback, I'll be responsive.

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@nkcsgexi @rudkx Okay to merge?

nkcsgexi commented:
The migrator part LGTM to merge!

rudkx commented:
Yes, merging.

rudkx commented:
@bjhomer It looks like this broke the foundation build: https://ci.swift.org/job/swift-PR-Linux-smoke-test/11232/consoleFull#-1189064102373bf607-d277-47c6-812e-27acc852624c

```
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift-corelibs-foundation/TestFoundation/TestURL.swift:38:38: error: cannot force unwrap value of non-optional type '[String : Any]'
13:46:22     guard let parsingTests = testRoot![kURLTestParsingTestsKey] as? [Any] else {
```

rudkx commented:
Okay, I don't see how this passed PR testing yesterday but would now fail today. I might be missing something that changed in foundation. Reverting for now, though.

bjhomer commented:
Fixed in https://github.com/apple/swift-corelibs-foundation/pull/1777
